### PR TITLE
Specialize `Base.similar` for `VectorOfArray` with multidimensional parent

### PR DESCRIPTION
**Pull request by jlchan**
This PR specializes `Base.similar` for `VectorOfArray` with a homogeneous (e.g., all elements are the same type) multidimensional parent array (introduced in https://github.com/SciML/RecursiveArrayTools.jl/pull/357). This ensures that `similar(VectorOfArray(array)).u` is similar to `array.u` where `array` is a homogeneous multi-dimensional array.

This PR also adds `Base.parent` for `VectorOfArray` (`parent(vec)` just returns `vec.u`)

### CODE REVIEW

**Comment by jlchan** [comment]
@ranocha I think with these changes, we can deal with multidimensional arrays of `SVector`

For example, the following works for me under this PR:
```julia
using StaticArrays
using RecursiveArrayTools
using OrdinaryDiffEq

function rhs!(duu::VectorOfArray, uu::VectorOfArray, p, t)
    du = parent(duu)
    u = parent(uu)
    du .= u
end

u = fill(SVector{2}(ones(2)), 2, 3)
ode = ODEProblem(rhs!, VectorOfArray(u), (0.0, 1.0))
sol = solve(ode, Tsit5())
```

File: src/vector_of_array.jl [hunk 1]
```diff
@@ -721,6 +724,18 @@ end
     VectorOfArray([similar(VA[:, i], T) for i in eachindex(VA.u)])
 end
 
+# for VectorOfArray with multi-dimensional parent arrays of arrays where all elements are the same type
+@inline function Base.similar(vec::VectorOfArray{
```
Comment (ChrisRackauckas) [inline src/vector_of_array.jl:728]:
```suggestion
function Base.similar(vec::VectorOfArray{
```

Is that well-tested? What's the reason for overriding the inlining calculation?

**Comment by ChrisRackauckas** [comment]
Add that as a regression test in the downstream portion (https://github.com/SciML/RecursiveArrayTools.jl/blob/master/test/upstream.jl which is for some reason named upstream we should fix that and move it to the other test group https://github.com/SciML/RecursiveArrayTools.jl/blob/master/test/runtests.jl#L63

File: src/vector_of_array.jl (hunk 1 above)
Comment (jlchan) [reply src/vector_of_array.jl]: Yeah, it's tested. 

I just copied the `@inline` from before, I didn't think about it overriding the other inlining.

Comment (jlchan) [reply src/vector_of_array.jl]: Thanks!

**Review by ChrisRackauckas — APPROVED** [review]

File: test/downstream/odesolve.jl [hunk 2]
```diff
@@ -49,3 +49,14 @@ end
...
         (0.0, 1.0)),
     Rodas5()).retcode == ReturnCode.Success
+
+function rhs!(duu::VectorOfArray, uu::VectorOfArray, p, t)
+    du = parent(duu)
+    u = parent(uu)
+    du .= u
+end
+
+u = fill(SVector{2}(ones(2)), 2, 3)
```
Comment (jlchan) [inline test/downstream/odesolve.jl:59]:
```suggestion
u = fill(ones(2), 2, 3)
```
Would need to add `StaticArrays` as dependency otherwise.

Comment (jlchan) [reply test/downstream/odesolve.jl:59]: Nvm, StaticArrays is already a dependency

Comment (ChrisRackauckas) [reply test/downstream/odesolve.jl:59]: ohh... I'll just revert the downstream dep

Comment (ChrisRackauckas) [reply test/downstream/odesolve.jl:59]: wait no, it's only a core dep, not a StaticArrays dep, so this is right.

**Pull request merged**